### PR TITLE
Separate FileHandle and BufferManagedFileHandle

### PR DESCRIPTION
**Pull request by ray6080**
This PR separates `FileHandle` and `BufferManagedFileHandle`.
The former one is used for files that are **not** managed by BM, and the latter one is for ones managed by BM.

- `BufferManagedFileHandle` supports both versioned and non-versioned file handles.
- Page resize logic now are all unified to grow one page group (1024 pages) at one time (grow means page capacity will increase 1024 at one time, not actual num of pages).
- `StorageStructureID` is moved from FileHandle to `StorageStructure`. `StorageStructureID` is mostly used to logically identify the semantic of a storage structure, for example, a node table column. FileHandle shouldn't be aware of those information.

### CODE REVIEW

File: src/include/storage/storage_structure/lists/list_headers.h [hunk 1]
```diff
@@ -124,7 +123,7 @@ class ListHeaders : public BaseListHeaders {
         return (*headersDiskArray)[offset];
     };
 
-    inline VersionedFileHandle* getFileHandle() { return versionedFileHandle.get(); }
+    inline FileHandle* getFileHandle() { return fileHandle.get(); }
```
Comment (andyfengHKU) [inline src/include/storage/storage_structure/lists/list_headers.h:126]: remove if not needed

File: src/include/storage/wal_replayer.h [hunk 2]
```diff
@@ -15,15 +14,16 @@ namespace storage {
...
     // This interface is used for recovery only. We always recover the disk files before
     // constructing the storageManager and catalog. So this specialized recovery constructor
     // doesn't take in storageManager and bufferManager.
-    WALReplayer(WAL* wal);
+    explicit WALReplayer(WAL* wal);
 
-    WALReplayer(WAL* wal, StorageManager* storageManager, BufferManager* bufferManager,
-        MemoryManager* memoryManager, catalog::Catalog* catalog, bool isCheckpoint);
+    WALReplayer(WAL* wal, StorageManager* storageManager, MemoryManager* memoryManager,
+        catalog::Catalog* catalog, bool isCheckpoint, bool isRecovering = false);
```
Comment (andyfengHKU) [inline src/include/storage/wal_replayer.h:26]: Unless you are setting isRecovering to True in any case, don't expose isRecovering as a default argument.

**Review by andyfengHKU — APPROVED** [review]

File: src/include/storage/storage_structure/lists/list_headers.h (hunk 1 above)
Comment (ray6080) [reply src/include/storage/storage_structure/lists/list_headers.h]: Done.

File: src/include/storage/wal_replayer.h (hunk 2 above)
Comment (ray6080) [reply src/include/storage/wal_replayer.h]: Done.

**Pull request merged**